src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs

```diff
@@ -0,0 +1,413 @@
+            var status = Registration.StreamSendDelegate(
+                _nativeObjPtr,
+                sendContext.Buffers,
+                sendContext.BufferCount,
+                (uint)flags,
+                sendContext.NativeClientSendContext);
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return sendContext.TcsTask; // TODO make QuicStream implement IValueTaskSource?
+        }
+
+        public Task StartAsync(QUIC_STREAM_START_FLAG flags)
+        {
+            var status = (uint)Registration.StreamStartDelegate(
+              _nativeObjPtr,
+              (uint)flags);
+            MsQuicStatusException.ThrowIfFailed(status);
+            return Task.CompletedTask;
+        }
+
+        public void ReceiveComplete(int bufferLength)
+        {
+            var status = (uint)Registration.StreamReceiveComplete(_nativeObjPtr, (ulong)bufferLength);
+            MsQuicStatusException.ThrowIfFailed(status);
+        }
+
+        public void ShutDown(
+            QUIC_STREAM_SHUTDOWN_FLAG flags,
+            ushort errorCode)
+        {
+            var status = (uint)Registration.StreamShutdownDelegate(
+                _nativeObjPtr,
+                (uint)flags,
+                errorCode);
+            MsQuicStatusException.ThrowIfFailed(status);
+        }
+
+        public void Close()
+        {
+            var status = (uint)Registration.StreamCloseDelegate?.Invoke(_nativeObjPtr);
+            MsQuicStatusException.ThrowIfFailed(status);
+        }
+
+        public long Handle { get => (long)_nativeObjPtr; }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+
+        public unsafe void EnableReceive()
+        {
+            var val = true;
+            var buffer = new QuicBuffer()
+            {
+                Length = sizeof(bool),
+                Buffer = (byte*)&val
+            };
+            SetParam(QUIC_PARAM_STREAM.RECEIVE_ENABLED, buffer);
+        }
+
+        private void SetParam(
+              QUIC_PARAM_STREAM param,
+              QuicBuffer buf)
+        {
+            MsQuicStatusException.ThrowIfFailed(Registration.UnsafeSetParam(
+                _nativeObjPtr,
+                (uint)QUIC_PARAM_LEVEL.SESSION,
+                (uint)param,
+                buf));
+        }
+
+
+        ~MsQuicStream()
+        {
+            Dispose(false);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_nativeObjPtr != IntPtr.Zero)
+            {
+                Registration.StreamCloseDelegate?.Invoke(_nativeObjPtr);
+            }
+
+            _nativeObjPtr = IntPtr.Zero;
+            Registration = null;
+
+            _disposed = true;
+        }
+    }
+
+    internal class FakeTlsConnectionFeature : ITlsConnectionFeature
```

Comment:
What's the point of this?

Reply:
Pretty much it's a feature that is checked up stack (Kestrel) to make sure the connection is "secure". Right now, we can't get this information from MsQuic (correct me if I'm wrong), so we have a fake one for now.